File: Excel/600-699/636/636 Challenge.py
import pandas as pd

path = "636 Repeat Customers in a Year.xlsx"
input = pd.read_excel(path, usecols="A:C", skiprows=1, nrows=88)
test = pd.read_excel(path,  usecols="E:G", skiprows=1, nrows=5)

input['Year'] = pd.DatetimeIndex(input['Date']).year

repeat_customers = (input.groupby(['Year', 'Customer', 'Store'])
                    .size()
                    .reset_index(name='n')
                    .query('n > 1')
                    .groupby('Year')
                    .agg(Count=('Customer', 'nunique'),
                         Customers=('Customer', lambda x: ', '.join(sorted(x.unique()))))
                    .reset_index())

print(all(test == repeat_customers)) # True

File: Excel/600-699/636/636 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/636 Repeat Customers in a Year.xlsx"
input = read_excel(path, range = "A2:C90")
test  = read_excel(path, range = "E2:G7")

repeat_customers = input %>%
  mutate(Year = year(Date)) %>%
  summarise(n = n(), .by = c(Year, Customer,Store)) %>%
  filter(n > 1) %>%
  summarise(
            Count = n_distinct(Customer),
            Customers = paste0(unique(sort(Customer)), collapse = ", "),
            .by = c(Year))

all.equal(repeat_customers, test)
#> [1] TRUE